python/code/classificationmethods/ManifoldDiscrimantAnalysisParafacTucker.py

```diff
@@ -0,0 +1,193 @@
+from abc import ABC, abstractmethod
+import numpy as np
+from scipy.stats import ortho_group  
+from pymanopt import Problem
+from pymanopt.manifolds import Product, Stiefel
+import tensorflow as tf
+from pymanopt.solvers import ConjugateGradient
+from sklearn.pipeline import Pipeline
+
+
+class ManifoldDiscrimantAnalysis(ABC, Pipeline):
+    def __init__(self, classes=None, opts=None, usestoppingcrit=True, maxits=1000,
+                 store={'Rw': None, 'Rb': None, 'QtRb': None, 'QtRw':  None, 'QtBQ': None, 'QtWQ': None, 'QtWQinvQtBQ': None},
+                 optmeth='ManOpt'):
+        self.classes = classes
+        self.opts = opts
+        self.usestoppingcrit = usestoppingcrit
+        self.maxits = maxits
+        self.store = store
+        self.optmeth = optmeth
+        if optmeth not in ['bo13', 'wen12', 'ManOpt']:
+            self.optmeth = 'ManOpt'
+            print("Chosen optimization method", optmeth, "has not been implemented")
+        self.Fdifftol = Fdifftol = 10**(-10)
+        self.Udifftol = Udifftol = 10**(-12)
+        self.MyCost = None
+        self.rotations = None
+
+
+    def set_tolerances(self, Fdifftol=10**(-10), Udifftol=10**(-12)):
+        self.Fdifftol = Fdifftol
+        self.Udifftol = Udifftol
+        
+    @abstractmethod 
+    def object_matrix_data(self, classmeandiffs, observationdiffs, nis, K1, K2, ): #This is a virtual function that any inheriting subclass should realize
+        return(None)
+
+
+    @abstractmethod
+    def my_cost(self): #, x, store, classmeandiffs, observationdiffs, nis, K1, K2):
+        pass
+        """
+        self.F = -np.linalg.trace(self.store['QtWQinvQtBQ'])
+        Q = tf.Variable(tf.placeholder(tf.float32))
+        """
+
+        #problem = Problem(manifold = manifold, cost = MyCost, arg = Q)
+
+        """
+        With tensorflow:
+            Q = tf.Variable(tf.placeholder(tf.float32))
+            self.MyCost = tf.linalg.trace(g(Q)) Operations to construct my cost inserted herein - potential to save a lot of calculations and definitions in the code when porting?
+                    
+            problem = Problem(manifold = manifold, cost = MyCost, arg = Q)
+            
+        """
+    @staticmethod
+    def class_based_differences(Xs, classes):
+        """
+        Calculate class based differences.
+
+        The purpose of this method is, given a labelled input data set to output the means of each class, and the data set recentered around the class means for each point.
```

Comment:
If we insert a line break here, we can keep the line length to within the PEP8 specifications.